{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import primer3 as pm"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CC</th>\n",
       "      <th>Seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;clonotype1_consensus_1</td>\n",
       "      <td>GGGAGACCTTGCCTATGGGACCATAGGATGCTAGCATACCCCTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;clonotype1_consensus_2</td>\n",
       "      <td>TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;clonotype2_consensus_1</td>\n",
       "      <td>TCTCGTGCTTACGTGGAGTTTCTATGAGTGAAGCCACTGCCTCATC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;clonotype2_consensus_2</td>\n",
       "      <td>GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;clonotype2_consensus_3</td>\n",
       "      <td>CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        CC                                                Seq\n",
       "0  >clonotype1_consensus_1  GGGAGACCTTGCCTATGGGACCATAGGATGCTAGCATACCCCTCCT...\n",
       "1  >clonotype1_consensus_2  TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...\n",
       "2  >clonotype2_consensus_1  TCTCGTGCTTACGTGGAGTTTCTATGAGTGAAGCCACTGCCTCATC...\n",
       "3  >clonotype2_consensus_2  GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...\n",
       "4  >clonotype2_consensus_3  CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT..."
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"inflammation_consensus.fasta.txt\", \"r\") as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "clonotypes = [line.strip() for i, line in enumerate(lines) if i % 2 == 0]\n",
    "seqs = [line.strip() for i, line in enumerate(lines) if i % 2 == 1]\n",
    "\n",
    "df = pd.DataFrame({\"CC\": clonotypes, \"Seq\": seqs})\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "alphas = []\n",
    "betas = []\n",
    "for seq in seqs:\n",
    "    if 'ACATCCAGAACCCAGAACCT' in seq:\n",
    "        alphas.append(seq)\n",
    "    else:\n",
    "        betas.append(seq)\n",
    "'''\n",
    "t = df['Seq'].str.contains('ACATCCAGAACCCAGAACC')\n",
    "alpha = df[t]\n",
    "beta = df[~t]\n",
    "\n",
    "alpha_orig = alpha.copy(deep = True)\n",
    "beta_orig = beta.copy(deep = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CAT'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rev = {'A':'T', 'C':'G', 'G':'C', 'T':'A'}\n",
    "def rev_comp(seq):\n",
    "    new_seq = \"\".join(map(rev.get, seq))  # Convert to string\n",
    "    return new_seq[::-1]\n",
    "rev_comp('ATG')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def no_repeat_neighbors(seq, index, direction):\n",
    "    \"\"\"\n",
    "    Returns True if the nucleotide at 'index' is NOT part of a contiguous block\n",
    "    of three or more of the same nucleotide. Returns False if it IS part of such a block.\n",
    "    \"\"\"\n",
    "    nucleotide = seq[index]\n",
    "    if nucleotide not in ('C', 'G'):\n",
    "        return True\n",
    "    \n",
    "    if direction == 'left':\n",
    "        if nucleotide == seq[index - 1] and seq[index - 1] == seq[index - 2] and seq[index - 2] == seq[index - 3]:\n",
    "            return False\n",
    "        \n",
    "    if direction == 'right':\n",
    "        if nucleotide == seq[index + 1] and seq[index + 1] == seq[index + 2] and seq[index + 2] == seq[index + 3]:\n",
    "            return False\n",
    "    \n",
    "    return True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'*': ['TGA', 'TAA', 'TAG'], 'A': ['GCC', 'GCT', 'GCA', 'GCG'], 'C': ['TGC', 'TGT'], 'D': ['GAC', 'GAT'], 'E': ['GAG', 'GAA'], 'F': ['TTC', 'TTT'], 'G': ['GGC', 'GGG', 'GGA', 'GGT'], 'H': ['CAC', 'CAT'], 'I': ['ATC', 'ATT', 'ATA'], 'K': ['AAG', 'AAA'], 'L': ['CTG', 'CTC', 'TTG', 'CTT', 'CTA', 'TTA'], 'M': ['ATG'], 'N': ['AAC', 'AAT'], 'P': ['CCC', 'CCT', 'CCA', 'CCG'], 'Q': ['CAG', 'CAA'], 'R': ['AGG', 'AGA', 'CGG', 'CGA', 'CGC', 'CGT'], 'S': ['AGC', 'TCC', 'TCT', 'AGT', 'TCA', 'TCG'], 'T': ['ACC', 'ACA', 'ACT', 'ACG'], 'V': ['GTG', 'GTC', 'GTT', 'GTA'], 'W': ['TGG'], 'Y': ['TAC', 'TAT']}\n"
     ]
    }
   ],
   "source": [
    "data = {\n",
    "    'Amino Acid': ['F', 'F', 'L', 'L', 'L', 'L', 'L', 'L', 'I', 'I', 'I', 'M', 'V', 'V', 'V', 'V', 'S', 'S', 'S', 'S', 'S', 'S', 'P', 'P', 'P', 'P', 'T', 'T', 'T', 'T', 'A', 'A', 'A', 'A', 'Y', 'Y', 'H', 'H', 'Q', 'Q', 'N', 'N', 'K', 'K', 'D', 'D', 'E', 'E', 'C', 'C', 'W', 'R', 'R', 'R', 'R', 'R', 'R', 'G', 'G', 'G', 'G', '*', '*', '*'],\n",
    "    'Triplet': ['TTC', 'TTT', 'CTG', 'CTC', 'CTA', 'TTG', 'TTA', 'CTT', 'ATC', 'ATT', 'ATA', 'ATG', 'GTG', 'GTC', 'GTT', 'GTA', 'AGC', 'AGT', 'TCC', 'TCT', 'TCA', 'TCG', 'CCC', 'CCT', 'CCA', 'CCG', 'ACC', 'ACT', 'ACA', 'ACG', 'GCC', 'GCT', 'GCA', 'GCG', 'TAC', 'TAT', 'CAC', 'CAT', 'CAG', 'CAA', 'AAC', 'AAT', 'AAG', 'AAA', 'GAC', 'GAT', 'GAG', 'GAA', 'TGC', 'TGT', 'TGG', 'AGG', 'AGA', 'CGG', 'CGA', 'CGC', 'CGT', 'GGC', 'GGT', 'GGG', 'GGA', 'TGA', 'TAA', 'TAG'],\n",
    "    'Frequency': [0.57, 0.43, 0.39, 0.20, 0.08, 0.13, 0.06, 0.13, 0.50, 0.34, 0.22, 1.00, 0.46, 0.25, 0.17, 0.12, 0.24, 0.15, 0.22, 0.19, 0.14, 0.05, 0.31, 0.30, 0.28, 0.10, 0.35, 0.25, 0.29, 0.11, 0.38, 0.29, 0.26, 0.10, 0.58, 0.43, 0.60, 0.40, 0.75, 0.25, 0.57, 0.43, 0.61, 0.39, 0.56, 0.44, 0.60, 0.26, 0.52, 0.48, 1.00, 0.22, 0.21, 0.19, 0.18, 0.12, 0.09, 0.33, 0.18, 0.26, 0.23, 0.52, 0.26, 0.22]\n",
    "}\n",
    "\n",
    "aas = pd.DataFrame(data)\n",
    "aas = aas.sort_values(by=['Amino Acid', 'Frequency'], ascending=[True, False])\n",
    "\n",
    "codon_dict = aas.groupby('Amino Acid')['Triplet'].apply(list).to_dict()\n",
    "print(codon_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Frequency doesn't matter, this is just one change. We just need to make the change and keep whatever nucleotide is there, the same\n",
    "def preventer(seq, index, direction):\n",
    "    if (no_repeat_neighbors(seq, index, direction)):\n",
    "        return True, seq\n",
    "    seq2 = ''\n",
    "    nucleotide = seq[index]\n",
    "    point = index % 3\n",
    "    if direction == 'left':\n",
    "        right = 2 - point + index\n",
    "        codons = [seq[right - 5:right - 2], seq[right - 2:right + 1]]\n",
    "        aa = [aas[aas['Triplet'] == codons[0]]['Amino Acid'].values[0], aas[aas['Triplet'] == codons[1]]['Amino Acid'].values[0]]\n",
    "        choices_1 = list(aas[aas['Amino Acid'] == aa[0]]['Triplet'])\n",
    "        choices_2 = list(aas[aas['Amino Acid'] == aa[1]]['Triplet'])\n",
    "        for choice in choices_2:\n",
    "            seq2 = seq[:right - 5] + choice + seq[right - 2:]\n",
    "            if no_repeat_neighbors(seq2, index, 'left'):\n",
    "                return True, seq2\n",
    "        for choice in choices_1:\n",
    "            seq2 = seq[:right - 2] + choice + seq[right+1:]\n",
    "            if seq2[index] != nucleotide:\n",
    "                continue\n",
    "            if no_repeat_neighbors(seq2, index, 'left'):\n",
    "                return True, seq2\n",
    "        \n",
    "    if direction == 'right':\n",
    "        right = 2 - point + index\n",
    "        codons = [seq[right + 1:right + 4], seq[right - 2:right + 1]]\n",
    "        aa = [aas[aas['Triplet'] == codons[0]]['Amino Acid'].values[0], aas[aas['Triplet'] == codons[1]]['Amino Acid'].values[0]]\n",
    "        choices_1 = list(aas[aas['Amino Acid'] == aa[0]]['Triplet'])\n",
    "        choices_2 = list(aas[aas['Amino Acid'] == aa[1]]['Triplet'])\n",
    "        for choice in choices_1:\n",
    "            seq2 = seq[:right+1] + choice + seq[right+4:]\n",
    "            if no_repeat_neighbors(seq2, index, 'right'):\n",
    "                return True, seq2\n",
    "        for choice in choices_2:\n",
    "            seq2 = seq[:right - 2] + choice + seq[right+1:]\n",
    "            if seq2[index] != nucleotide:\n",
    "                continue\n",
    "            if no_repeat_neighbors(seq2, index, 'right'):\n",
    "                return True, seq2\n",
    "    return False, None\n",
    "            \n",
    "\n",
    "\n",
    "t = preventer('TGAGTCCCCGATGCCCCTGC', 7, 'left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"Missing_Alphas.txt\", \"w\") as f:\n",
    "    f.write(\"Unable to find working sequences for clonotypes: \\n\")\n",
    "\n",
    "with open(\"Missing_Betas.txt\", \"w\") as f:\n",
    "    f.write(\"Unable to find working sequences for clonotypes: \\n\")\n",
    "\n",
    "def missing(tp, clono, reason):\n",
    "    with open(f\"Missing_{tp}.txt\", \"a\") as f:\n",
    "        f.write(f\"{clono}: {reason}\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TRAV"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CC</th>\n",
       "      <th>Seq</th>\n",
       "      <th>TRAV_start_ind</th>\n",
       "      <th>TRAV_end_ind</th>\n",
       "      <th>TRAV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;clonotype1_consensus_2</td>\n",
       "      <td>TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>ATGAAGAGGCTGCTGTGTTCTCTGCTGGGGCTTCTGTGCACCCAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;clonotype2_consensus_2</td>\n",
       "      <td>GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...</td>\n",
       "      <td>238.0</td>\n",
       "      <td>580.0</td>\n",
       "      <td>ATGAAAACATATGCTCCTACATTATTCATGTTTCTATGGCTGCAGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;clonotype2_consensus_3</td>\n",
       "      <td>CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>380.0</td>\n",
       "      <td>ATGCATTCCTTACATGTTTCACTAGTGTTCCTCTGGCTTCAACTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>&gt;clonotype3_consensus_2</td>\n",
       "      <td>GAACTGCATTTCCTGCTGGGGTGGTTACACCACACAGACCAGTTAC...</td>\n",
       "      <td>135.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>ATGCTGATTCTAAGCCTGTTGGGAGCAGCCTTTGGCTCCATTTGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>&gt;clonotype4_consensus_2</td>\n",
       "      <td>TGGGGGAGGTCAGGGAGAGTCTCTTTTCTAGTTACCAACTTAGAGT...</td>\n",
       "      <td>85.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>ATGAAATCCTTTAGTATTTCCCTAGTGGTCCTGTGGCTTCAGCTAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21129</th>\n",
       "      <td>&gt;clonotype9965_consensus_2</td>\n",
       "      <td>AGAGAGCTGAGGTGCCTGTGAAGTTGGAGTCAGTGTTCTGTGGGAG...</td>\n",
       "      <td>89.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>ATGGACAAGATCCTGACAGCATCGTTTTTACTTCTAGGCCTTCACC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21130</th>\n",
       "      <td>&gt;clonotype9966_consensus_1</td>\n",
       "      <td>GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21131</th>\n",
       "      <td>&gt;clonotype9967_consensus_1</td>\n",
       "      <td>GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21132</th>\n",
       "      <td>&gt;clonotype9968_consensus_1</td>\n",
       "      <td>GGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATG...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21133</th>\n",
       "      <td>&gt;clonotype9969_consensus_1</td>\n",
       "      <td>TGGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAG...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>383.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10522 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               CC  \\\n",
       "1         >clonotype1_consensus_2   \n",
       "3         >clonotype2_consensus_2   \n",
       "4         >clonotype2_consensus_3   \n",
       "6         >clonotype3_consensus_2   \n",
       "8         >clonotype4_consensus_2   \n",
       "...                           ...   \n",
       "21129  >clonotype9965_consensus_2   \n",
       "21130  >clonotype9966_consensus_1   \n",
       "21131  >clonotype9967_consensus_1   \n",
       "21132  >clonotype9968_consensus_1   \n",
       "21133  >clonotype9969_consensus_1   \n",
       "\n",
       "                                                     Seq  TRAV_start_ind  \\\n",
       "1      TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...            59.0   \n",
       "3      GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...           238.0   \n",
       "4      CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT...            44.0   \n",
       "6      GAACTGCATTTCCTGCTGGGGTGGTTACACCACACAGACCAGTTAC...           135.0   \n",
       "8      TGGGGGAGGTCAGGGAGAGTCTCTTTTCTAGTTACCAACTTAGAGT...            85.0   \n",
       "...                                                  ...             ...   \n",
       "21129  AGAGAGCTGAGGTGCCTGTGAAGTTGGAGTCAGTGTTCTGTGGGAG...            89.0   \n",
       "21130  GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...             6.0   \n",
       "21131  GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...             6.0   \n",
       "21132  GGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATG...             4.0   \n",
       "21133  TGGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAG...             7.0   \n",
       "\n",
       "       TRAV_end_ind                                               TRAV  \n",
       "1             387.0  ATGAAGAGGCTGCTGTGTTCTCTGCTGGGGCTTCTGTGCACCCAGG...  \n",
       "3             580.0  ATGAAAACATATGCTCCTACATTATTCATGTTTCTATGGCTGCAGC...  \n",
       "4             380.0  ATGCATTCCTTACATGTTTCACTAGTGTTCCTCTGGCTTCAACTAG...  \n",
       "6             483.0  ATGCTGATTCTAAGCCTGTTGGGAGCAGCCTTTGGCTCCATTTGTT...  \n",
       "8             417.0  ATGAAATCCTTTAGTATTTCCCTAGTGGTCCTGTGGCTTCAGCTAA...  \n",
       "...             ...                                                ...  \n",
       "21129         447.0  ATGGACAAGATCCTGACAGCATCGTTTTTACTTCTAGGCCTTCACC...  \n",
       "21130         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...  \n",
       "21131         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...  \n",
       "21132         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...  \n",
       "21133         383.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...  \n",
       "\n",
       "[10522 rows x 5 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"TRAVs.fasta.txt\", \"r\") as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "name = [line.strip() for i, line in enumerate(lines) if i % 2 == 0]\n",
    "travs = [line.strip() for i, line in enumerate(lines) if i % 2 == 1]\n",
    "\n",
    "def TRAV(seq, CC):\n",
    "    end_ind = None\n",
    "    trav_poss = []\n",
    "    for trav in travs:\n",
    "        if trav[:30] in seq: #Going to assume we need up to 100 nucleotides of trav to find the unique one it corresponds to (CHECK THIS)\n",
    "            trav_poss.append(trav)\n",
    "            \n",
    "    if len(trav_poss) == 0:\n",
    "        missing('Alphas', CC, 'Could not find TRAV that worked')\n",
    "        return None, None, None\n",
    "\n",
    "    start_ind = seq.find(trav_poss[0][:30])\n",
    "    temp_inds = [len(trav) for trav in trav_poss]\n",
    "    while temp_inds[0] > 200:\n",
    "        for ind in range(len(trav_poss)):\n",
    "            if seq[start_ind:start_ind + temp_inds[ind]] != trav_poss[ind][:temp_inds[ind]]:\n",
    "                temp_inds[ind]-=1\n",
    "            else:\n",
    "                end_ind = start_ind + temp_inds[ind]\n",
    "                return start_ind, end_ind, seq[start_ind:end_ind]\n",
    "    if end_ind == None:\n",
    "        missing('Alphas', CC, 'Could not find TRAV Primer that worked')\n",
    "        return None, None, None\n",
    "    return start_ind, end_ind, seq[start_ind:end_ind]\n",
    "\n",
    "alpha = alpha.copy()\n",
    "#alpha[['TRAV_start_ind', 'TRAV_end_ind', 'TRAV']] = alpha['Seq'].apply(lambda seq: pd.Series(TRAV(seq)))\n",
    "alpha[['TRAV_start_ind', 'TRAV_end_ind', 'TRAV']] = alpha.apply(lambda row: pd.Series(TRAV(row['Seq'], row['CC'])), axis = 1)\n",
    "#alpha = alpha.dropna()\n",
    "alpha\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CDR3a Reverse Primer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CC</th>\n",
       "      <th>Seq</th>\n",
       "      <th>TRAV_start_ind</th>\n",
       "      <th>TRAV_end_ind</th>\n",
       "      <th>TRAV</th>\n",
       "      <th>CDR3a_overlap_start_ind</th>\n",
       "      <th>CDR3a_overlap_end_ind</th>\n",
       "      <th>CDR3a_overlap_seq</th>\n",
       "      <th>CDR3a_overlap_Tm</th>\n",
       "      <th>CDR3a_Rev_Primer</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;clonotype1_consensus_2</td>\n",
       "      <td>TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>ATGAAGAGGCTGCTGTGTTCTCTGCTGGGGCTTCTGTGCACCCAGG...</td>\n",
       "      <td>410.0</td>\n",
       "      <td>430.0</td>\n",
       "      <td>CTCATCTTTGGACTGGGGAC</td>\n",
       "      <td>55.485372</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;clonotype2_consensus_2</td>\n",
       "      <td>GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...</td>\n",
       "      <td>238.0</td>\n",
       "      <td>580.0</td>\n",
       "      <td>ATGAAAACATATGCTCCTACATTATTCATGTTTCTATGGCTGCAGC...</td>\n",
       "      <td>605.0</td>\n",
       "      <td>622.0</td>\n",
       "      <td>CCTTTGGGGATGGGACC</td>\n",
       "      <td>54.257077</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGCTTCACTGTGAGCACGGTCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;clonotype2_consensus_3</td>\n",
       "      <td>CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>380.0</td>\n",
       "      <td>ATGCATTCCTTACATGTTTCACTAGTGTTCCTCTGGCTTCAACTAG...</td>\n",
       "      <td>404.0</td>\n",
       "      <td>424.0</td>\n",
       "      <td>CTCATCTTTGGACTGGGGAC</td>\n",
       "      <td>55.485372</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>&gt;clonotype3_consensus_2</td>\n",
       "      <td>GAACTGCATTTCCTGCTGGGGTGGTTACACCACACAGACCAGTTAC...</td>\n",
       "      <td>135.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>ATGCTGATTCTAAGCCTGTTGGGAGCAGCCTTTGGCTCCATTTGTT...</td>\n",
       "      <td>503.0</td>\n",
       "      <td>523.0</td>\n",
       "      <td>CTTCTTTGGTGATGGGACGC</td>\n",
       "      <td>57.090331</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTGGGCTTCACCACCAGCTGCGTCCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>&gt;clonotype4_consensus_2</td>\n",
       "      <td>TGGGGGAGGTCAGGGAGAGTCTCTTTTCTAGTTACCAACTTAGAGT...</td>\n",
       "      <td>85.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>ATGAAATCCTTTAGTATTTCCCTAGTGGTCCTGTGGCTTCAGCTAA...</td>\n",
       "      <td>439.0</td>\n",
       "      <td>459.0</td>\n",
       "      <td>CTTTACTTCGGATCTGGCAC</td>\n",
       "      <td>54.117413</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGCTCTACAGTGAGTTTGGTGCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21129</th>\n",
       "      <td>&gt;clonotype9965_consensus_2</td>\n",
       "      <td>AGAGAGCTGAGGTGCCTGTGAAGTTGGAGTCAGTGTTCTGTGGGAG...</td>\n",
       "      <td>89.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>ATGGACAAGATCCTGACAGCATCGTTTTTACTTCTAGGCCTTCACC...</td>\n",
       "      <td>470.0</td>\n",
       "      <td>492.0</td>\n",
       "      <td>CTGACATTTGGGAAAGGAACTC</td>\n",
       "      <td>55.087404</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGAGTCACAGTTAAGAGAGTTCCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21130</th>\n",
       "      <td>&gt;clonotype9966_consensus_1</td>\n",
       "      <td>GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>408.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>CTCACGTTTGGACACGGC</td>\n",
       "      <td>56.981512</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21131</th>\n",
       "      <td>&gt;clonotype9967_consensus_1</td>\n",
       "      <td>GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>408.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>CTCACGTTTGGACACGGC</td>\n",
       "      <td>56.981512</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21132</th>\n",
       "      <td>&gt;clonotype9968_consensus_1</td>\n",
       "      <td>GGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATG...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>406.0</td>\n",
       "      <td>424.0</td>\n",
       "      <td>CTCACGTTTGGACACGGC</td>\n",
       "      <td>56.981512</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21133</th>\n",
       "      <td>&gt;clonotype9969_consensus_1</td>\n",
       "      <td>TGGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAG...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>383.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>409.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>CTCATCTTTGGAATTGGGACTC</td>\n",
       "      <td>54.617867</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGCTTGACAGAAAGCAGAGTCCCA...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10522 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               CC  \\\n",
       "1         >clonotype1_consensus_2   \n",
       "3         >clonotype2_consensus_2   \n",
       "4         >clonotype2_consensus_3   \n",
       "6         >clonotype3_consensus_2   \n",
       "8         >clonotype4_consensus_2   \n",
       "...                           ...   \n",
       "21129  >clonotype9965_consensus_2   \n",
       "21130  >clonotype9966_consensus_1   \n",
       "21131  >clonotype9967_consensus_1   \n",
       "21132  >clonotype9968_consensus_1   \n",
       "21133  >clonotype9969_consensus_1   \n",
       "\n",
       "                                                     Seq  TRAV_start_ind  \\\n",
       "1      TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...            59.0   \n",
       "3      GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...           238.0   \n",
       "4      CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT...            44.0   \n",
       "6      GAACTGCATTTCCTGCTGGGGTGGTTACACCACACAGACCAGTTAC...           135.0   \n",
       "8      TGGGGGAGGTCAGGGAGAGTCTCTTTTCTAGTTACCAACTTAGAGT...            85.0   \n",
       "...                                                  ...             ...   \n",
       "21129  AGAGAGCTGAGGTGCCTGTGAAGTTGGAGTCAGTGTTCTGTGGGAG...            89.0   \n",
       "21130  GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...             6.0   \n",
       "21131  GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...             6.0   \n",
       "21132  GGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATG...             4.0   \n",
       "21133  TGGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAG...             7.0   \n",
       "\n",
       "       TRAV_end_ind                                               TRAV  \\\n",
       "1             387.0  ATGAAGAGGCTGCTGTGTTCTCTGCTGGGGCTTCTGTGCACCCAGG...   \n",
       "3             580.0  ATGAAAACATATGCTCCTACATTATTCATGTTTCTATGGCTGCAGC...   \n",
       "4             380.0  ATGCATTCCTTACATGTTTCACTAGTGTTCCTCTGGCTTCAACTAG...   \n",
       "6             483.0  ATGCTGATTCTAAGCCTGTTGGGAGCAGCCTTTGGCTCCATTTGTT...   \n",
       "8             417.0  ATGAAATCCTTTAGTATTTCCCTAGTGGTCCTGTGGCTTCAGCTAA...   \n",
       "...             ...                                                ...   \n",
       "21129         447.0  ATGGACAAGATCCTGACAGCATCGTTTTTACTTCTAGGCCTTCACC...   \n",
       "21130         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "21131         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "21132         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "21133         383.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "\n",
       "       CDR3a_overlap_start_ind  CDR3a_overlap_end_ind       CDR3a_overlap_seq  \\\n",
       "1                        410.0                  430.0    CTCATCTTTGGACTGGGGAC   \n",
       "3                        605.0                  622.0       CCTTTGGGGATGGGACC   \n",
       "4                        404.0                  424.0    CTCATCTTTGGACTGGGGAC   \n",
       "6                        503.0                  523.0    CTTCTTTGGTGATGGGACGC   \n",
       "8                        439.0                  459.0    CTTTACTTCGGATCTGGCAC   \n",
       "...                        ...                    ...                     ...   \n",
       "21129                    470.0                  492.0  CTGACATTTGGGAAAGGAACTC   \n",
       "21130                    408.0                  426.0      CTCACGTTTGGACACGGC   \n",
       "21131                    408.0                  426.0      CTCACGTTTGGACACGGC   \n",
       "21132                    406.0                  424.0      CTCACGTTTGGACACGGC   \n",
       "21133                    409.0                  431.0  CTCATCTTTGGAATTGGGACTC   \n",
       "\n",
       "       CDR3a_overlap_Tm                                   CDR3a_Rev_Primer  \n",
       "1             55.485372  AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...  \n",
       "3             54.257077  AGGTTCTGGGTTCTGGATGTTTGGCTTCACTGTGAGCACGGTCCCA...  \n",
       "4             55.485372  AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...  \n",
       "6             57.090331  AGGTTCTGGGTTCTGGATGTTGGGCTTCACCACCAGCTGCGTCCCA...  \n",
       "8             54.117413  AGGTTCTGGGTTCTGGATGTTTGGCTCTACAGTGAGTTTGGTGCCA...  \n",
       "...                 ...                                                ...  \n",
       "21129         55.087404  AGGTTCTGGGTTCTGGATGTTTGGAGTCACAGTTAAGAGAGTTCCT...  \n",
       "21130         56.981512  AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...  \n",
       "21131         56.981512  AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...  \n",
       "21132         56.981512  AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...  \n",
       "21133         54.617867  AGGTTCTGGGTTCTGGATGTTTGGCTTGACAGAAAGCAGAGTCCCA...  \n",
       "\n",
       "[10522 rows x 10 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def CDR_Reverse(seq, CC):\n",
    "    ind = seq.find('ACATCCAGAACCCAGAACCT')\n",
    "    rev_prime_ending = ind + 20\n",
    "    ind = ind - 40\n",
    "    point = ind + min(seq[ind:].find('C'), seq[ind:].find('G'))\n",
    "    end_point = None\n",
    "    rev_primer = None\n",
    "    for end in range(point + 17, point + 31):\n",
    "        if (seq[end - 1] == 'C' or seq[end - 1] == 'G') and (pm.calc_tm(seq[point:end]) >= 54 and pm.calc_tm(seq[point:end]) <= 58):\n",
    "            found, temp_seq = preventer(seq, end - 1, 'left')\n",
    "            if found:\n",
    "                seq = temp_seq\n",
    "            else:\n",
    "                continue\n",
    "            end_point = end\n",
    "            rev_primer = seq[point:rev_prime_ending]\n",
    "            break\n",
    "    if end_point == None or rev_primer == None:\n",
    "        missing('Alphas', CC, 'Could not find CDR Reverse Primer that worked')\n",
    "        return None, None, None, None\n",
    "    return point, end_point, seq[point:end_point], pm.calc_tm(seq[point:end_point]), rev_comp(rev_primer), \n",
    "\n",
    "alpha = alpha.copy()\n",
    "#alpha[['CDR3a_overlap_start_ind', 'CDR3a_overlap_end_ind', 'CDR3a_overlap_seq', 'CDR3a_overlap_Tm', 'CDR3a_Rev_Primer']] = alpha['Seq'].apply(lambda seq: pd.Series(CDR_Reverse(seq)))\n",
    "alpha[['CDR3a_overlap_start_ind', 'CDR3a_overlap_end_ind', 'CDR3a_overlap_seq', 'CDR3a_overlap_Tm', 'CDR3a_Rev_Primer']] = alpha.apply(lambda row: pd.Series(CDR_Reverse(row['Seq'], row['CC'])), axis = 1)\n",
    "#alpha = alpha.dropna()\n",
    "alpha  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CDR3a forward and forward outer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CC</th>\n",
       "      <th>Seq</th>\n",
       "      <th>TRAV_start_ind</th>\n",
       "      <th>TRAV_end_ind</th>\n",
       "      <th>TRAV</th>\n",
       "      <th>CDR3a_overlap_start_ind</th>\n",
       "      <th>CDR3a_overlap_end_ind</th>\n",
       "      <th>CDR3a_overlap_seq</th>\n",
       "      <th>CDR3a_overlap_Tm</th>\n",
       "      <th>CDR3a_Rev_Primer</th>\n",
       "      <th>TRAV_overlap_start_ind</th>\n",
       "      <th>TRAV_overlap_end_ind</th>\n",
       "      <th>TRAV_overlap_seq</th>\n",
       "      <th>CDR3a_forward</th>\n",
       "      <th>CDR3a_forward_outer</th>\n",
       "      <th>CDR3a_forward_outer_Tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>&gt;clonotype1_consensus_2</td>\n",
       "      <td>TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...</td>\n",
       "      <td>59.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>ATGAAGAGGCTGCTGTGTTCTCTGCTGGGGCTTCTGTGCACCCAGG...</td>\n",
       "      <td>410.0</td>\n",
       "      <td>430.0</td>\n",
       "      <td>CTCATCTTTGGACTGGGGAC</td>\n",
       "      <td>55.485372</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...</td>\n",
       "      <td>366.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>GCACTTATTTCTGTGCTATAG</td>\n",
       "      <td>TTATTTCTGTGCTATAGGGGGAACAGGCAATACCGGAAAACTCATC...</td>\n",
       "      <td>GCACTTATTTCTGTGCTATAGGGGGAAC</td>\n",
       "      <td>55.909889</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>&gt;clonotype2_consensus_2</td>\n",
       "      <td>GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...</td>\n",
       "      <td>238.0</td>\n",
       "      <td>580.0</td>\n",
       "      <td>ATGAAAACATATGCTCCTACATTATTCATGTTTCTATGGCTGCAGC...</td>\n",
       "      <td>605.0</td>\n",
       "      <td>622.0</td>\n",
       "      <td>CCTTTGGGGATGGGACC</td>\n",
       "      <td>54.257077</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGCTTCACTGTGAGCACGGTCCCA...</td>\n",
       "      <td>564.0</td>\n",
       "      <td>580.0</td>\n",
       "      <td>CTTCTGTGCTGCAAGT</td>\n",
       "      <td>CTTCTGTGCTGCAAGTCCGAGCACCAATACAGGCAAATTAACCTTT...</td>\n",
       "      <td>CTTCTGTGCTGCAAGTCCG</td>\n",
       "      <td>57.040949</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>&gt;clonotype2_consensus_3</td>\n",
       "      <td>CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT...</td>\n",
       "      <td>44.0</td>\n",
       "      <td>380.0</td>\n",
       "      <td>ATGCATTCCTTACATGTTTCACTAGTGTTCCTCTGGCTTCAACTAG...</td>\n",
       "      <td>404.0</td>\n",
       "      <td>424.0</td>\n",
       "      <td>CTCATCTTTGGACTGGGGAC</td>\n",
       "      <td>55.485372</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...</td>\n",
       "      <td>364.0</td>\n",
       "      <td>380.0</td>\n",
       "      <td>CCTCTGTGCAGTGAGC</td>\n",
       "      <td>CCTCTGTGCAGTGAGCCGTATAACAGGCAATACCGGAAAACTCATC...</td>\n",
       "      <td>CCTCTGTGCAGTGAGCC</td>\n",
       "      <td>55.143497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>&gt;clonotype3_consensus_2</td>\n",
       "      <td>GAACTGCATTTCCTGCTGGGGTGGTTACACCACACAGACCAGTTAC...</td>\n",
       "      <td>135.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>ATGCTGATTCTAAGCCTGTTGGGAGCAGCCTTTGGCTCCATTTGTT...</td>\n",
       "      <td>503.0</td>\n",
       "      <td>523.0</td>\n",
       "      <td>CTTCTTTGGTGATGGGACGC</td>\n",
       "      <td>57.090331</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTGGGCTTCACCACCAGCTGCGTCCCA...</td>\n",
       "      <td>461.0</td>\n",
       "      <td>483.0</td>\n",
       "      <td>AGTATATTTCTGTGCTATGAGA</td>\n",
       "      <td>TATATTTCTGTGCTATGAGATGGAATAGCAATAACAGAATCTTCTT...</td>\n",
       "      <td>AGTATATTTCTGTGCTATGAGATGGAATAG</td>\n",
       "      <td>54.839399</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>&gt;clonotype4_consensus_2</td>\n",
       "      <td>TGGGGGAGGTCAGGGAGAGTCTCTTTTCTAGTTACCAACTTAGAGT...</td>\n",
       "      <td>85.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>ATGAAATCCTTTAGTATTTCCCTAGTGGTCCTGTGGCTTCAGCTAA...</td>\n",
       "      <td>439.0</td>\n",
       "      <td>459.0</td>\n",
       "      <td>CTTTACTTCGGATCTGGCAC</td>\n",
       "      <td>54.117413</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGCTCTACAGTGAGTTTGGTGCCA...</td>\n",
       "      <td>401.0</td>\n",
       "      <td>417.0</td>\n",
       "      <td>TCTACCTCTGTGCAGC</td>\n",
       "      <td>TCTACCTCTGTGCAGCCCAGGGGTCTAATTACAACGTGCTTTACTT...</td>\n",
       "      <td>TCTACCTCTGTGCAGCCC</td>\n",
       "      <td>56.203817</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21129</th>\n",
       "      <td>&gt;clonotype9965_consensus_2</td>\n",
       "      <td>AGAGAGCTGAGGTGCCTGTGAAGTTGGAGTCAGTGTTCTGTGGGAG...</td>\n",
       "      <td>89.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>ATGGACAAGATCCTGACAGCATCGTTTTTACTTCTAGGCCTTCACC...</td>\n",
       "      <td>470.0</td>\n",
       "      <td>492.0</td>\n",
       "      <td>CTGACATTTGGGAAAGGAACTC</td>\n",
       "      <td>55.087404</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGAGTCACAGTTAAGAGAGTTCCT...</td>\n",
       "      <td>430.0</td>\n",
       "      <td>447.0</td>\n",
       "      <td>CTACTTCTGTGCAGCAA</td>\n",
       "      <td>ACTTCTGTGCAGCAAACAGTGGAGGCAGCAATTACAAACTGACATT...</td>\n",
       "      <td>CTACTTCTGTGCAGCAAACAG</td>\n",
       "      <td>54.236044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21130</th>\n",
       "      <td>&gt;clonotype9966_consensus_1</td>\n",
       "      <td>GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>408.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>CTCACGTTTGGACACGGC</td>\n",
       "      <td>56.981512</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...</td>\n",
       "      <td>363.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>TACTACTGTGCTCTGAGT</td>\n",
       "      <td>TACTGTGCTCTGAGTACTAACACTGGAGCTAACACTGGAAAGCTCA...</td>\n",
       "      <td>TACTACTGTGCTCTGAGTACTAACAC</td>\n",
       "      <td>55.229221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21131</th>\n",
       "      <td>&gt;clonotype9967_consensus_1</td>\n",
       "      <td>GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>408.0</td>\n",
       "      <td>426.0</td>\n",
       "      <td>CTCACGTTTGGACACGGC</td>\n",
       "      <td>56.981512</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...</td>\n",
       "      <td>363.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>TACTACTGTGCTCTGAGT</td>\n",
       "      <td>TACTGTGCTCTGAGTACTAACACTGGAGCTAACACTGGAAAGCTCA...</td>\n",
       "      <td>TACTACTGTGCTCTGAGTACTAACAC</td>\n",
       "      <td>55.229221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21132</th>\n",
       "      <td>&gt;clonotype9968_consensus_1</td>\n",
       "      <td>GGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATG...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>406.0</td>\n",
       "      <td>424.0</td>\n",
       "      <td>CTCACGTTTGGACACGGC</td>\n",
       "      <td>56.981512</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...</td>\n",
       "      <td>363.0</td>\n",
       "      <td>381.0</td>\n",
       "      <td>CTACTGTGCTCTGAGTGA</td>\n",
       "      <td>TACTGTGCTCTGAGTGATCGAACTGGAGCTAACACTGGAAAGCTCA...</td>\n",
       "      <td>CTACTGTGCTCTGAGTGATCG</td>\n",
       "      <td>54.663627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21133</th>\n",
       "      <td>&gt;clonotype9969_consensus_1</td>\n",
       "      <td>TGGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAG...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>383.0</td>\n",
       "      <td>ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...</td>\n",
       "      <td>409.0</td>\n",
       "      <td>431.0</td>\n",
       "      <td>CTCATCTTTGGAATTGGGACTC</td>\n",
       "      <td>54.617867</td>\n",
       "      <td>AGGTTCTGGGTTCTGGATGTTTGGCTTGACAGAAAGCAGAGTCCCA...</td>\n",
       "      <td>365.0</td>\n",
       "      <td>383.0</td>\n",
       "      <td>ACTACTGTGCTCTGAGTG</td>\n",
       "      <td>GTGCTCTGAGTGTGAATTATGGGAGCAGTGGCAACAAGCTCATCTT...</td>\n",
       "      <td>ACTACTGTGCTCTGAGTGTGAATTATGG</td>\n",
       "      <td>54.988734</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10522 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               CC  \\\n",
       "1         >clonotype1_consensus_2   \n",
       "3         >clonotype2_consensus_2   \n",
       "4         >clonotype2_consensus_3   \n",
       "6         >clonotype3_consensus_2   \n",
       "8         >clonotype4_consensus_2   \n",
       "...                           ...   \n",
       "21129  >clonotype9965_consensus_2   \n",
       "21130  >clonotype9966_consensus_1   \n",
       "21131  >clonotype9967_consensus_1   \n",
       "21132  >clonotype9968_consensus_1   \n",
       "21133  >clonotype9969_consensus_1   \n",
       "\n",
       "                                                     Seq  TRAV_start_ind  \\\n",
       "1      TGACACAACAGCCCAGGGACTGGTTACTTGCTTCTGTCTCCAGCAT...            59.0   \n",
       "3      GGGTTAGGAATGCAACCACAAGGTGTCAGGAGATCTTAACTGCTGG...           238.0   \n",
       "4      CTCCTTTTGCTGGCTTGAAGTGTGAATCTTCAGTGAAAAGAGTAAT...            44.0   \n",
       "6      GAACTGCATTTCCTGCTGGGGTGGTTACACCACACAGACCAGTTAC...           135.0   \n",
       "8      TGGGGGAGGTCAGGGAGAGTCTCTTTTCTAGTTACCAACTTAGAGT...            85.0   \n",
       "...                                                  ...             ...   \n",
       "21129  AGAGAGCTGAGGTGCCTGTGAAGTTGGAGTCAGTGTTCTGTGGGAG...            89.0   \n",
       "21130  GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...             6.0   \n",
       "21131  GGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGA...             6.0   \n",
       "21132  GGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATG...             4.0   \n",
       "21133  TGGGGACATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAG...             7.0   \n",
       "\n",
       "       TRAV_end_ind                                               TRAV  \\\n",
       "1             387.0  ATGAAGAGGCTGCTGTGTTCTCTGCTGGGGCTTCTGTGCACCCAGG...   \n",
       "3             580.0  ATGAAAACATATGCTCCTACATTATTCATGTTTCTATGGCTGCAGC...   \n",
       "4             380.0  ATGCATTCCTTACATGTTTCACTAGTGTTCCTCTGGCTTCAACTAG...   \n",
       "6             483.0  ATGCTGATTCTAAGCCTGTTGGGAGCAGCCTTTGGCTCCATTTGTT...   \n",
       "8             417.0  ATGAAATCCTTTAGTATTTCCCTAGTGGTCCTGTGGCTTCAGCTAA...   \n",
       "...             ...                                                ...   \n",
       "21129         447.0  ATGGACAAGATCCTGACAGCATCGTTTTTACTTCTAGGCCTTCACC...   \n",
       "21130         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "21131         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "21132         381.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "21133         383.0  ATGCTAAGAATCATGAACACTGTCGAGATGGGTCTAAAGATGAACT...   \n",
       "\n",
       "       CDR3a_overlap_start_ind  CDR3a_overlap_end_ind       CDR3a_overlap_seq  \\\n",
       "1                        410.0                  430.0    CTCATCTTTGGACTGGGGAC   \n",
       "3                        605.0                  622.0       CCTTTGGGGATGGGACC   \n",
       "4                        404.0                  424.0    CTCATCTTTGGACTGGGGAC   \n",
       "6                        503.0                  523.0    CTTCTTTGGTGATGGGACGC   \n",
       "8                        439.0                  459.0    CTTTACTTCGGATCTGGCAC   \n",
       "...                        ...                    ...                     ...   \n",
       "21129                    470.0                  492.0  CTGACATTTGGGAAAGGAACTC   \n",
       "21130                    408.0                  426.0      CTCACGTTTGGACACGGC   \n",
       "21131                    408.0                  426.0      CTCACGTTTGGACACGGC   \n",
       "21132                    406.0                  424.0      CTCACGTTTGGACACGGC   \n",
       "21133                    409.0                  431.0  CTCATCTTTGGAATTGGGACTC   \n",
       "\n",
       "       CDR3a_overlap_Tm                                   CDR3a_Rev_Primer  \\\n",
       "1             55.485372  AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...   \n",
       "3             54.257077  AGGTTCTGGGTTCTGGATGTTTGGCTTCACTGTGAGCACGGTCCCA...   \n",
       "4             55.485372  AGGTTCTGGGTTCTGGATGTCTGGTTGCACTTGTAAAGTTGTCCCC...   \n",
       "6             57.090331  AGGTTCTGGGTTCTGGATGTTGGGCTTCACCACCAGCTGCGTCCCA...   \n",
       "8             54.117413  AGGTTCTGGGTTCTGGATGTTTGGCTCTACAGTGAGTTTGGTGCCA...   \n",
       "...                 ...                                                ...   \n",
       "21129         55.087404  AGGTTCTGGGTTCTGGATGTTTGGAGTCACAGTTAAGAGAGTTCCT...   \n",
       "21130         56.981512  AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...   \n",
       "21131         56.981512  AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...   \n",
       "21132         56.981512  AGGTTCTGGGTTCTGGATGTTTGGATGGACCCTAAGGATGGTGCCG...   \n",
       "21133         54.617867  AGGTTCTGGGTTCTGGATGTTTGGCTTGACAGAAAGCAGAGTCCCA...   \n",
       "\n",
       "       TRAV_overlap_start_ind  TRAV_overlap_end_ind        TRAV_overlap_seq  \\\n",
       "1                       366.0                 387.0   GCACTTATTTCTGTGCTATAG   \n",
       "3                       564.0                 580.0        CTTCTGTGCTGCAAGT   \n",
       "4                       364.0                 380.0        CCTCTGTGCAGTGAGC   \n",
       "6                       461.0                 483.0  AGTATATTTCTGTGCTATGAGA   \n",
       "8                       401.0                 417.0        TCTACCTCTGTGCAGC   \n",
       "...                       ...                   ...                     ...   \n",
       "21129                   430.0                 447.0       CTACTTCTGTGCAGCAA   \n",
       "21130                   363.0                 381.0      TACTACTGTGCTCTGAGT   \n",
       "21131                   363.0                 381.0      TACTACTGTGCTCTGAGT   \n",
       "21132                   363.0                 381.0      CTACTGTGCTCTGAGTGA   \n",
       "21133                   365.0                 383.0      ACTACTGTGCTCTGAGTG   \n",
       "\n",
       "                                           CDR3a_forward  \\\n",
       "1      TTATTTCTGTGCTATAGGGGGAACAGGCAATACCGGAAAACTCATC...   \n",
       "3      CTTCTGTGCTGCAAGTCCGAGCACCAATACAGGCAAATTAACCTTT...   \n",
       "4      CCTCTGTGCAGTGAGCCGTATAACAGGCAATACCGGAAAACTCATC...   \n",
       "6      TATATTTCTGTGCTATGAGATGGAATAGCAATAACAGAATCTTCTT...   \n",
       "8      TCTACCTCTGTGCAGCCCAGGGGTCTAATTACAACGTGCTTTACTT...   \n",
       "...                                                  ...   \n",
       "21129  ACTTCTGTGCAGCAAACAGTGGAGGCAGCAATTACAAACTGACATT...   \n",
       "21130  TACTGTGCTCTGAGTACTAACACTGGAGCTAACACTGGAAAGCTCA...   \n",
       "21131  TACTGTGCTCTGAGTACTAACACTGGAGCTAACACTGGAAAGCTCA...   \n",
       "21132  TACTGTGCTCTGAGTGATCGAACTGGAGCTAACACTGGAAAGCTCA...   \n",
       "21133  GTGCTCTGAGTGTGAATTATGGGAGCAGTGGCAACAAGCTCATCTT...   \n",
       "\n",
       "                  CDR3a_forward_outer  CDR3a_forward_outer_Tm  \n",
       "1        GCACTTATTTCTGTGCTATAGGGGGAAC               55.909889  \n",
       "3                 CTTCTGTGCTGCAAGTCCG               57.040949  \n",
       "4                   CCTCTGTGCAGTGAGCC               55.143497  \n",
       "6      AGTATATTTCTGTGCTATGAGATGGAATAG               54.839399  \n",
       "8                  TCTACCTCTGTGCAGCCC               56.203817  \n",
       "...                               ...                     ...  \n",
       "21129           CTACTTCTGTGCAGCAAACAG               54.236044  \n",
       "21130      TACTACTGTGCTCTGAGTACTAACAC               55.229221  \n",
       "21131      TACTACTGTGCTCTGAGTACTAACAC               55.229221  \n",
       "21132           CTACTGTGCTCTGAGTGATCG               54.663627  \n",
       "21133    ACTACTGTGCTCTGAGTGTGAATTATGG               54.988734  \n",
       "\n",
       "[10522 rows x 16 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def CDRA_Forwards(seq, trav_overlap_end, cdr_overlap_end, CC):\n",
    "    if pd.isna(trav_overlap_end) or pd.isna(cdr_overlap_end):\n",
    "        return None, None, None, None, None, None\n",
    "    trav_overlap_end, cdr_overlap_end = int(trav_overlap_end), int(cdr_overlap_end)\n",
    "    trav_overlap_beg = trav_overlap_end - 16\n",
    "    while pm.calc_tm(seq[trav_overlap_beg:trav_overlap_end]) < 49:\n",
    "        trav_overlap_beg-=1\n",
    "        if trav_overlap_beg < 0:\n",
    "            missing('Alphas', CC, 'Could not find TRAV Overlap that worked')\n",
    "            return None, None, None, None, None, None\n",
    "    cdr_forward_beg = trav_overlap_beg\n",
    "    while cdr_overlap_end - cdr_forward_beg > 60 or no_repeat_neighbors(seq, cdr_forward_beg, 'right') == False:\n",
    "        cdr_forward_beg+=1\n",
    "        if cdr_forward_beg >= cdr_overlap_end:\n",
    "            missing('Alphas', CC, 'Could not find CDRA Forward Primer that worked')\n",
    "            return trav_overlap_beg, trav_overlap_end, seq[trav_overlap_beg:trav_overlap_end], None, None, None\n",
    "    cdr_forward = seq[cdr_forward_beg:cdr_overlap_end]\n",
    "    cdr_forward_outer_end_ind = trav_overlap_beg + 16\n",
    "    #while pm.calc_tm(seq[trav_overlap_beg:cdr_forward_outer_end_ind]) < 49 or (pm.calc_tm(seq[cdr_forward_beg:cdr_forward_outer_end_ind]) < 54 or pm.calc_tm(seq[cdr_forward_beg:cdr_forward_outer_end_ind]) > 58) or (seq[cdr_forward_outer_end_ind - 1] not in ('C','G')): #or no_repeat_neighbors(seq, cdr_forward_outer_end_ind - 1, 'left') == False:\n",
    "    while True:\n",
    "        if pm.calc_tm(seq[trav_overlap_beg:cdr_forward_outer_end_ind]) >= 49 and (pm.calc_tm(seq[cdr_forward_beg:cdr_forward_outer_end_ind]) >= 54 and pm.calc_tm(seq[cdr_forward_beg:cdr_forward_outer_end_ind]) <= 58) and (seq[cdr_forward_outer_end_ind - 1] in ('C','G')):\n",
    "            found, temp_seq = preventer(seq, cdr_forward_outer_end_ind - 1, 'left')\n",
    "            if found:\n",
    "                seq = temp_seq\n",
    "                break\n",
    "        cdr_forward_outer_end_ind+=1\n",
    "        if cdr_forward_outer_end_ind > cdr_overlap_end:\n",
    "            missing('Alphas', CC, 'Could not find CDRA Forward Outer Primer that worked')\n",
    "            return trav_overlap_beg, trav_overlap_end, seq[trav_overlap_beg:trav_overlap_end], cdr_forward, None, None\n",
    "    cdr_forward_outer = seq[trav_overlap_beg:cdr_forward_outer_end_ind]\n",
    "    \n",
    "    temp_cdr = cdr_forward_outer\n",
    "    while temp_cdr not in cdr_forward:\n",
    "        temp_cdr = temp_cdr[1:]\n",
    "\n",
    "    return trav_overlap_beg, trav_overlap_end, seq[trav_overlap_beg:trav_overlap_end], cdr_forward, cdr_forward_outer, pm.calc_tm(temp_cdr)\n",
    "    \n",
    "alpha = alpha.copy()\n",
    "alpha[['TRAV_overlap_start_ind', 'TRAV_overlap_end_ind', 'TRAV_overlap_seq', 'CDR3a_forward', 'CDR3a_forward_outer', 'CDR3a_forward_outer_Tm']] = alpha.apply(\n",
    "    lambda row: pd.Series(CDRA_Forwards(row['Seq'], row['TRAV_end_ind'], row['CDR3a_overlap_end_ind'], row['CC'])),\n",
    "    axis=1\n",
    ")\n",
    "#alpha = alpha.dropna()\n",
    "alpha  "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "CDR3b"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CC</th>\n",
       "      <th>Seq</th>\n",
       "      <th>TRBV_overlap_start_ind</th>\n",
       "      <th>TRBV_overlap_end_ind</th>\n",
       "      <th>TRBV_overlap_seq</th>\n",
       "      <th>Prim_overlap_start_ind</th>\n",
       "      <th>Prim_overlap_end_ind</th>\n",
       "      <th>Primer_overlap_seq</th>\n",
       "      <th>Primer_overlap_Tm</th>\n",
       "      <th>CDR3b_forward</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;clonotype1_consensus_1</td>\n",
       "      <td>GGGAGACCTTGCCTATGGGACCATAGGATGCTAGCATACCCCTCCT...</td>\n",
       "      <td>436.0</td>\n",
       "      <td>452.0</td>\n",
       "      <td>TGCCAGCAGCTTAGCC</td>\n",
       "      <td>457.0</td>\n",
       "      <td>476.0</td>\n",
       "      <td>GGACTCCGACTACACCTTC</td>\n",
       "      <td>54.489626</td>\n",
       "      <td>TGCCAGCAGCTTAGCCGGGGGGGACTCCGACTACACCTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;clonotype2_consensus_1</td>\n",
       "      <td>TCTCGTGCTTACGTGGAGTTTCTATGAGTGAAGCCACTGCCTCATC...</td>\n",
       "      <td>374.0</td>\n",
       "      <td>394.0</td>\n",
       "      <td>GTATCTTTGTGCAAGCAGCT</td>\n",
       "      <td>396.0</td>\n",
       "      <td>414.0</td>\n",
       "      <td>GGCGGATGGAACACCTTG</td>\n",
       "      <td>56.326643</td>\n",
       "      <td>GTATCTTTGTGCAAGCAGCTCTGGCGGATGGAACACCTTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>&gt;clonotype3_consensus_1</td>\n",
       "      <td>AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...</td>\n",
       "      <td>352.0</td>\n",
       "      <td>368.0</td>\n",
       "      <td>GTGCCAGCAGCCAAGA</td>\n",
       "      <td>366.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>GATAACAATTCTGGAAATACGCTC</td>\n",
       "      <td>54.076533</td>\n",
       "      <td>GTGCCAGCAGCCAAGATAACAATTCTGGAAATACGCTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>&gt;clonotype4_consensus_1</td>\n",
       "      <td>GAAAACTGGCTTGCTCATGTAGAGAAGTGGTGGAGTGTCTTAACTG...</td>\n",
       "      <td>476.0</td>\n",
       "      <td>494.0</td>\n",
       "      <td>TTCTGTGCCAGCAGTGAT</td>\n",
       "      <td>491.0</td>\n",
       "      <td>516.0</td>\n",
       "      <td>GATAATAGGGGAGAAGTCTTCTTTG</td>\n",
       "      <td>54.571719</td>\n",
       "      <td>TTCTGTGCCAGCAGTGATAATAGGGGAGAAGTCTTCTTTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>&gt;clonotype5_consensus_1</td>\n",
       "      <td>AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...</td>\n",
       "      <td>350.0</td>\n",
       "      <td>366.0</td>\n",
       "      <td>TTGTGCCAGCAGCCAA</td>\n",
       "      <td>369.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>GGGCAAGACACCCAGTAC</td>\n",
       "      <td>55.201463</td>\n",
       "      <td>TTGTGCCAGCAGCCAACTGGGGCAAGACACCCAGTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21119</th>\n",
       "      <td>&gt;clonotype9960_consensus_1</td>\n",
       "      <td>TGGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATG...</td>\n",
       "      <td>400.0</td>\n",
       "      <td>416.0</td>\n",
       "      <td>TGTGCCAGCGGTGATG</td>\n",
       "      <td>419.0</td>\n",
       "      <td>439.0</td>\n",
       "      <td>GCTGGGGGTTTAACCAAGAC</td>\n",
       "      <td>56.641294</td>\n",
       "      <td>TGTGCCAGCGGTGATGCGGGCTGGGGGTTTAACCAAGAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21121</th>\n",
       "      <td>&gt;clonotype9961_consensus_1</td>\n",
       "      <td>GGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTC...</td>\n",
       "      <td>395.0</td>\n",
       "      <td>411.0</td>\n",
       "      <td>TTCTGTGCCAGCGGTG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21124</th>\n",
       "      <td>&gt;clonotype9963_consensus_1</td>\n",
       "      <td>GAAGCAGGCTGAGCCTAAAGCACAGAAGGGCATAGCCAACACTGTA...</td>\n",
       "      <td>543.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>TGTACTTCTGTGCCAGCTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21126</th>\n",
       "      <td>&gt;clonotype9964_consensus_1</td>\n",
       "      <td>TGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGT...</td>\n",
       "      <td>399.0</td>\n",
       "      <td>415.0</td>\n",
       "      <td>TGTGCCAGCGGTGATG</td>\n",
       "      <td>419.0</td>\n",
       "      <td>436.0</td>\n",
       "      <td>GGGACTGGGGGGTCATC</td>\n",
       "      <td>56.161153</td>\n",
       "      <td>TGTGCCAGCGGTGATGCAATGGGACTGGGGGGTCATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21128</th>\n",
       "      <td>&gt;clonotype9965_consensus_1</td>\n",
       "      <td>GGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTCTA...</td>\n",
       "      <td>396.0</td>\n",
       "      <td>412.0</td>\n",
       "      <td>TGTGCCAGCGGTGATG</td>\n",
       "      <td>415.0</td>\n",
       "      <td>436.0</td>\n",
       "      <td>GGACTGGGGTTTTTAACCAAG</td>\n",
       "      <td>54.846900</td>\n",
       "      <td>TGTGCCAGCGGTGATGCAGGGACTGGGGTTTTTAACCAAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10612 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               CC  \\\n",
       "0         >clonotype1_consensus_1   \n",
       "2         >clonotype2_consensus_1   \n",
       "5         >clonotype3_consensus_1   \n",
       "7         >clonotype4_consensus_1   \n",
       "9         >clonotype5_consensus_1   \n",
       "...                           ...   \n",
       "21119  >clonotype9960_consensus_1   \n",
       "21121  >clonotype9961_consensus_1   \n",
       "21124  >clonotype9963_consensus_1   \n",
       "21126  >clonotype9964_consensus_1   \n",
       "21128  >clonotype9965_consensus_1   \n",
       "\n",
       "                                                     Seq  \\\n",
       "0      GGGAGACCTTGCCTATGGGACCATAGGATGCTAGCATACCCCTCCT...   \n",
       "2      TCTCGTGCTTACGTGGAGTTTCTATGAGTGAAGCCACTGCCTCATC...   \n",
       "5      AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...   \n",
       "7      GAAAACTGGCTTGCTCATGTAGAGAAGTGGTGGAGTGTCTTAACTG...   \n",
       "9      AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...   \n",
       "...                                                  ...   \n",
       "21119  TGGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATG...   \n",
       "21121  GGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTC...   \n",
       "21124  GAAGCAGGCTGAGCCTAAAGCACAGAAGGGCATAGCCAACACTGTA...   \n",
       "21126  TGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGT...   \n",
       "21128  GGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTCTA...   \n",
       "\n",
       "       TRBV_overlap_start_ind  TRBV_overlap_end_ind      TRBV_overlap_seq  \\\n",
       "0                       436.0                 452.0      TGCCAGCAGCTTAGCC   \n",
       "2                       374.0                 394.0  GTATCTTTGTGCAAGCAGCT   \n",
       "5                       352.0                 368.0      GTGCCAGCAGCCAAGA   \n",
       "7                       476.0                 494.0    TTCTGTGCCAGCAGTGAT   \n",
       "9                       350.0                 366.0      TTGTGCCAGCAGCCAA   \n",
       "...                       ...                   ...                   ...   \n",
       "21119                   400.0                 416.0      TGTGCCAGCGGTGATG   \n",
       "21121                   395.0                 411.0      TTCTGTGCCAGCGGTG   \n",
       "21124                   543.0                 562.0   TGTACTTCTGTGCCAGCTC   \n",
       "21126                   399.0                 415.0      TGTGCCAGCGGTGATG   \n",
       "21128                   396.0                 412.0      TGTGCCAGCGGTGATG   \n",
       "\n",
       "       Prim_overlap_start_ind  Prim_overlap_end_ind  \\\n",
       "0                       457.0                 476.0   \n",
       "2                       396.0                 414.0   \n",
       "5                       366.0                 390.0   \n",
       "7                       491.0                 516.0   \n",
       "9                       369.0                 387.0   \n",
       "...                       ...                   ...   \n",
       "21119                   419.0                 439.0   \n",
       "21121                     NaN                   NaN   \n",
       "21124                     NaN                   NaN   \n",
       "21126                   419.0                 436.0   \n",
       "21128                   415.0                 436.0   \n",
       "\n",
       "              Primer_overlap_seq  Primer_overlap_Tm  \\\n",
       "0            GGACTCCGACTACACCTTC          54.489626   \n",
       "2             GGCGGATGGAACACCTTG          56.326643   \n",
       "5       GATAACAATTCTGGAAATACGCTC          54.076533   \n",
       "7      GATAATAGGGGAGAAGTCTTCTTTG          54.571719   \n",
       "9             GGGCAAGACACCCAGTAC          55.201463   \n",
       "...                          ...                ...   \n",
       "21119       GCTGGGGGTTTAACCAAGAC          56.641294   \n",
       "21121                       None                NaN   \n",
       "21124                       None                NaN   \n",
       "21126          GGGACTGGGGGGTCATC          56.161153   \n",
       "21128      GGACTGGGGTTTTTAACCAAG          54.846900   \n",
       "\n",
       "                                  CDR3b_forward  \n",
       "0      TGCCAGCAGCTTAGCCGGGGGGGACTCCGACTACACCTTC  \n",
       "2      GTATCTTTGTGCAAGCAGCTCTGGCGGATGGAACACCTTG  \n",
       "5        GTGCCAGCAGCCAAGATAACAATTCTGGAAATACGCTC  \n",
       "7      TTCTGTGCCAGCAGTGATAATAGGGGAGAAGTCTTCTTTG  \n",
       "9         TTGTGCCAGCAGCCAACTGGGGCAAGACACCCAGTAC  \n",
       "...                                         ...  \n",
       "21119   TGTGCCAGCGGTGATGCGGGCTGGGGGTTTAACCAAGAC  \n",
       "21121                                      None  \n",
       "21124                                      None  \n",
       "21126     TGTGCCAGCGGTGATGCAATGGGACTGGGGGGTCATC  \n",
       "21128  TGTGCCAGCGGTGATGCAGGGACTGGGGTTTTTAACCAAG  \n",
       "\n",
       "[10612 rows x 10 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "with open(\"TRBVs.fasta.txt\", \"r\") as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "name = [line.strip() for i, line in enumerate(lines) if i % 2 == 0]\n",
    "trbvs = [line.strip() for i, line in enumerate(lines) if i % 2 == 1]\n",
    "missing_num = []\n",
    "\n",
    "def CDRB(seq, CC):\n",
    "    cdrb_start_ind = 0\n",
    "    for trbv in trbvs:\n",
    "        trbv_start_ind = seq.find(trbv[:80])\n",
    "        temp_ind = len(trbv)\n",
    "        if trbv_start_ind != -1:\n",
    "            while seq[trbv_start_ind:trbv_start_ind + temp_ind] != trbv[:temp_ind]:\n",
    "                temp_ind-=1\n",
    "                if temp_ind < 200:\n",
    "                    missing('Betas', CC, 'Could not find TRBV Sequence that worked') \n",
    "                    return None, None, None, None, None, None, None, None\n",
    "            cdrb_start_ind = trbv_start_ind + temp_ind\n",
    "            break\n",
    "    trbv_overlap_end_ind = cdrb_start_ind\n",
    "    trbv_overlap_start_ind = cdrb_start_ind\n",
    "    while pm.calc_tm(seq[trbv_overlap_start_ind:trbv_overlap_end_ind]) < 54 or pm.calc_tm(seq[trbv_overlap_start_ind:trbv_overlap_end_ind]) > 58:\n",
    "        trbv_overlap_start_ind-=1\n",
    "        if trbv_overlap_start_ind < trbv_overlap_end_ind - 100:\n",
    "            missing('Betas', CC, 'Could not find TRBV Overlap Primer that worked')\n",
    "            return None, None, None, None, None, None, None, None\n",
    "\n",
    "    cdrb_end_ind = trbv_overlap_start_ind + 40\n",
    "    #while ((seq[cdrb_end_ind - 1] not in ('C', 'G')) or (cdrb_end_ind - cdrb_start_ind > 60)): #or (not no_repeat_neighbors(seq, cdrb_end_ind - 1, 'left'))):\n",
    "    while True:\n",
    "        if ((seq[cdrb_end_ind - 1] in ('C', 'G')) and (cdrb_end_ind - cdrb_start_ind <= 60)):\n",
    "            found, temp_seq = preventer(seq, cdrb_end_ind - 1, 'left')\n",
    "            if found:\n",
    "                seq = temp_seq\n",
    "                break\n",
    "        cdrb_end_ind-=1\n",
    "        if cdrb_end_ind < trbv_overlap_start_ind:\n",
    "            missing('Betas', CC, 'Could not find CDR3b Primer that worked')\n",
    "            return trbv_overlap_start_ind, trbv_overlap_end_ind, seq[trbv_overlap_start_ind:trbv_overlap_end_ind], None, None, None, None, None\n",
    "\n",
    "    prim_overlap_end_ind = cdrb_end_ind\n",
    "    prim_overlap_start_ind = prim_overlap_end_ind - 16\n",
    "    #while pm.calc_tm(seq[prim_overlap_start_ind:cdrb_end_ind]) < 54 or pm.calc_tm(seq[prim_overlap_start_ind:cdrb_end_ind]) > 58 or seq[prim_overlap_start_ind] not in ('C', 'G'): #or (not no_repeat_neighbors(seq, prim_overlap_start_ind, 'right')):\n",
    "    while True:\n",
    "        if pm.calc_tm(seq[prim_overlap_start_ind:cdrb_end_ind]) >= 54 and pm.calc_tm(seq[prim_overlap_start_ind:cdrb_end_ind]) <= 58 and seq[prim_overlap_start_ind] in ('C', 'G'):\n",
    "            found, temp_seq = preventer(seq, prim_overlap_start_ind, 'right')\n",
    "            if found:\n",
    "                seq = temp_seq\n",
    "                break\n",
    "        prim_overlap_start_ind-=1\n",
    "        if prim_overlap_end_ind - prim_overlap_start_ind > 30:\n",
    "            missing('Betas', CC, 'Could not find Primer Overlap that worked')\n",
    "            return trbv_overlap_start_ind, trbv_overlap_end_ind, seq[trbv_overlap_start_ind:trbv_overlap_end_ind], None, None, None, None, None\n",
    "\n",
    "    return trbv_overlap_start_ind, trbv_overlap_end_ind, seq[trbv_overlap_start_ind:trbv_overlap_end_ind], prim_overlap_start_ind, prim_overlap_end_ind, seq[prim_overlap_start_ind:prim_overlap_end_ind], pm.calc_tm(seq[prim_overlap_start_ind:prim_overlap_end_ind]), seq[trbv_overlap_start_ind:cdrb_end_ind]#seq[cdrb_start_ind:cdrb_end_ind]\n",
    "\n",
    "beta = beta.copy()\n",
    "#beta[['TRBV_overlap_start_ind', 'TRBV_overlap_end_ind', 'TRBV_overlap_seq', 'Prim_overlap_start_ind', 'Prim_overlap_end_ind', 'Primer_overlap_seq', 'Primer_overlap_Tm', 'CDR3b_forward']] = beta['Seq'].apply(lambda seq: pd.Series(CDRB(seq)))\n",
    "beta[['TRBV_overlap_start_ind', 'TRBV_overlap_end_ind', 'TRBV_overlap_seq', 'Prim_overlap_start_ind', 'Prim_overlap_end_ind', 'Primer_overlap_seq', 'Primer_overlap_Tm', 'CDR3b_forward']] = beta.apply(lambda row: pd.Series(CDRB(row['Seq'], row['CC'])), axis = 1)\n",
    "#beta = beta.dropna()\n",
    "beta  \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CC</th>\n",
       "      <th>Seq</th>\n",
       "      <th>TRBV_overlap_start_ind</th>\n",
       "      <th>TRBV_overlap_end_ind</th>\n",
       "      <th>TRBV_overlap_seq</th>\n",
       "      <th>Prim_overlap_start_ind</th>\n",
       "      <th>Prim_overlap_end_ind</th>\n",
       "      <th>Primer_overlap_seq</th>\n",
       "      <th>Primer_overlap_Tm</th>\n",
       "      <th>CDR3b_forward</th>\n",
       "      <th>TRBC_overlap_start_ind</th>\n",
       "      <th>TRBC_overlap_end_ind</th>\n",
       "      <th>TRBC_overlap_seq</th>\n",
       "      <th>CDR3B_Reverse_Primer</th>\n",
       "      <th>CDR3B_Outer_Reverse_Primer</th>\n",
       "      <th>CDR3B_rev_overlap_Tm</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>&gt;clonotype1_consensus_1</td>\n",
       "      <td>GGGAGACCTTGCCTATGGGACCATAGGATGCTAGCATACCCCTCCT...</td>\n",
       "      <td>436.0</td>\n",
       "      <td>452.0</td>\n",
       "      <td>TGCCAGCAGCTTAGCC</td>\n",
       "      <td>457.0</td>\n",
       "      <td>476.0</td>\n",
       "      <td>GGACTCCGACTACACCTTC</td>\n",
       "      <td>54.489626</td>\n",
       "      <td>TGCCAGCAGCTTAGCCGGGGGGGACTCCGACTACACCTTC</td>\n",
       "      <td>504.0</td>\n",
       "      <td>525.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>TTTCTCAGATCCTCTATTACCAAAAGCCTGGTCCCTGAGCCGAAGG...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTATTACCAAAAG</td>\n",
       "      <td>54.568380</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>&gt;clonotype2_consensus_1</td>\n",
       "      <td>TCTCGTGCTTACGTGGAGTTTCTATGAGTGAAGCCACTGCCTCATC...</td>\n",
       "      <td>374.0</td>\n",
       "      <td>394.0</td>\n",
       "      <td>GTATCTTTGTGCAAGCAGCT</td>\n",
       "      <td>396.0</td>\n",
       "      <td>414.0</td>\n",
       "      <td>GGCGGATGGAACACCTTG</td>\n",
       "      <td>56.326643</td>\n",
       "      <td>GTATCTTTGTGCAAGCAGCTCTGGCGGATGGAACACCTTG</td>\n",
       "      <td>448.0</td>\n",
       "      <td>469.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>CAGATCCTCTAGCACCGATAGTCGGGTGCCCGCACCAAAGTACAAG...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTAGCACCGATAG</td>\n",
       "      <td>54.765029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>&gt;clonotype3_consensus_1</td>\n",
       "      <td>AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...</td>\n",
       "      <td>352.0</td>\n",
       "      <td>368.0</td>\n",
       "      <td>GTGCCAGCAGCCAAGA</td>\n",
       "      <td>366.0</td>\n",
       "      <td>390.0</td>\n",
       "      <td>GATAACAATTCTGGAAATACGCTC</td>\n",
       "      <td>54.076533</td>\n",
       "      <td>GTGCCAGCAGCCAAGATAACAATTCTGGAAATACGCTC</td>\n",
       "      <td>424.0</td>\n",
       "      <td>445.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>CTCTACAACAATGAGCCGGCTTCCTTCTCCAAAATAGAGCGTATTT...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTACAACAATGAGCCGGC</td>\n",
       "      <td>56.885491</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>&gt;clonotype4_consensus_1</td>\n",
       "      <td>GAAAACTGGCTTGCTCATGTAGAGAAGTGGTGGAGTGTCTTAACTG...</td>\n",
       "      <td>476.0</td>\n",
       "      <td>494.0</td>\n",
       "      <td>TTCTGTGCCAGCAGTGAT</td>\n",
       "      <td>491.0</td>\n",
       "      <td>516.0</td>\n",
       "      <td>GATAATAGGGGAGAAGTCTTCTTTG</td>\n",
       "      <td>54.571719</td>\n",
       "      <td>TTCTGTGCCAGCAGTGATAATAGGGGAGAAGTCTTCTTTG</td>\n",
       "      <td>543.0</td>\n",
       "      <td>564.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>CAGATCCTCTACAACTGTGAGTCTGGTTCCTTTACCAAAGAAGACT...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTACAACTGTGAGT</td>\n",
       "      <td>54.796917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>&gt;clonotype5_consensus_1</td>\n",
       "      <td>AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...</td>\n",
       "      <td>350.0</td>\n",
       "      <td>366.0</td>\n",
       "      <td>TTGTGCCAGCAGCCAA</td>\n",
       "      <td>369.0</td>\n",
       "      <td>387.0</td>\n",
       "      <td>GGGCAAGACACCCAGTAC</td>\n",
       "      <td>55.201463</td>\n",
       "      <td>TTGTGCCAGCAGCCAACTGGGGCAAGACACCCAGTAC</td>\n",
       "      <td>418.0</td>\n",
       "      <td>439.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>TCTCAGATCCTCTAACACGAGGAGCCGAGTGCCTGGCCCAAAGTAC...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTAACACGAGG</td>\n",
       "      <td>56.330476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21119</th>\n",
       "      <td>&gt;clonotype9960_consensus_1</td>\n",
       "      <td>TGGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATG...</td>\n",
       "      <td>400.0</td>\n",
       "      <td>416.0</td>\n",
       "      <td>TGTGCCAGCGGTGATG</td>\n",
       "      <td>419.0</td>\n",
       "      <td>439.0</td>\n",
       "      <td>GCTGGGGGTTTAACCAAGAC</td>\n",
       "      <td>56.641294</td>\n",
       "      <td>TGTGCCAGCGGTGATGCGGGCTGGGGGTTTAACCAAGAC</td>\n",
       "      <td>479.0</td>\n",
       "      <td>500.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>CTAACACGAGGAGCCGAGTGCCTGGCCCAAAGTACTGGGTGTCTTG...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTAACACGAGGAGCCGAG</td>\n",
       "      <td>54.812132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21121</th>\n",
       "      <td>&gt;clonotype9961_consensus_1</td>\n",
       "      <td>GGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTC...</td>\n",
       "      <td>395.0</td>\n",
       "      <td>411.0</td>\n",
       "      <td>TTCTGTGCCAGCGGTG</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21124</th>\n",
       "      <td>&gt;clonotype9963_consensus_1</td>\n",
       "      <td>GAAGCAGGCTGAGCCTAAAGCACAGAAGGGCATAGCCAACACTGTA...</td>\n",
       "      <td>543.0</td>\n",
       "      <td>562.0</td>\n",
       "      <td>TGTACTTCTGTGCCAGCTC</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21126</th>\n",
       "      <td>&gt;clonotype9964_consensus_1</td>\n",
       "      <td>TGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGT...</td>\n",
       "      <td>399.0</td>\n",
       "      <td>415.0</td>\n",
       "      <td>TGTGCCAGCGGTGATG</td>\n",
       "      <td>419.0</td>\n",
       "      <td>436.0</td>\n",
       "      <td>GGGACTGGGGGGTCATC</td>\n",
       "      <td>56.161153</td>\n",
       "      <td>TGTGCCAGCGGTGATGCAATGGGACTGGGGGGTCATC</td>\n",
       "      <td>481.0</td>\n",
       "      <td>502.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>AGCACCGATAGTCGGGTGCCCGCACCAAAGTACAAGGTGTTTTGAT...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTAGCACCGATAGTCGGGT</td>\n",
       "      <td>55.047754</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21128</th>\n",
       "      <td>&gt;clonotype9965_consensus_1</td>\n",
       "      <td>GGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTCTA...</td>\n",
       "      <td>396.0</td>\n",
       "      <td>412.0</td>\n",
       "      <td>TGTGCCAGCGGTGATG</td>\n",
       "      <td>415.0</td>\n",
       "      <td>436.0</td>\n",
       "      <td>GGACTGGGGTTTTTAACCAAG</td>\n",
       "      <td>54.846900</td>\n",
       "      <td>TGTGCCAGCGGTGATGCAGGGACTGGGGTTTTTAACCAAG</td>\n",
       "      <td>478.0</td>\n",
       "      <td>499.0</td>\n",
       "      <td>AGGATCTGAGAAATGTGACTC</td>\n",
       "      <td>ACACGAGGAGCCGAGTGCCTGGCCCAAAGTACTGGGTGTCTTGGTT...</td>\n",
       "      <td>GAGTCACATTTCTCAGATCCTCTAACACGAGGAGCCGAGT</td>\n",
       "      <td>54.787622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>10612 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               CC  \\\n",
       "0         >clonotype1_consensus_1   \n",
       "2         >clonotype2_consensus_1   \n",
       "5         >clonotype3_consensus_1   \n",
       "7         >clonotype4_consensus_1   \n",
       "9         >clonotype5_consensus_1   \n",
       "...                           ...   \n",
       "21119  >clonotype9960_consensus_1   \n",
       "21121  >clonotype9961_consensus_1   \n",
       "21124  >clonotype9963_consensus_1   \n",
       "21126  >clonotype9964_consensus_1   \n",
       "21128  >clonotype9965_consensus_1   \n",
       "\n",
       "                                                     Seq  \\\n",
       "0      GGGAGACCTTGCCTATGGGACCATAGGATGCTAGCATACCCCTCCT...   \n",
       "2      TCTCGTGCTTACGTGGAGTTTCTATGAGTGAAGCCACTGCCTCATC...   \n",
       "5      AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...   \n",
       "7      GAAAACTGGCTTGCTCATGTAGAGAAGTGGTGGAGTGTCTTAACTG...   \n",
       "9      AATACCCGTCTGGAGCCTGATTCCACCATGAGCTGCAGGCTTCTCC...   \n",
       "...                                                  ...   \n",
       "21119  TGGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATG...   \n",
       "21121  GGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTC...   \n",
       "21124  GAAGCAGGCTGAGCCTAAAGCACAGAAGGGCATAGCCAACACTGTA...   \n",
       "21126  TGGGGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGT...   \n",
       "21128  GGCTCTCTCACCAAAGAGACCAGTATCCTGAGAGGAAGCATGTCTA...   \n",
       "\n",
       "       TRBV_overlap_start_ind  TRBV_overlap_end_ind      TRBV_overlap_seq  \\\n",
       "0                       436.0                 452.0      TGCCAGCAGCTTAGCC   \n",
       "2                       374.0                 394.0  GTATCTTTGTGCAAGCAGCT   \n",
       "5                       352.0                 368.0      GTGCCAGCAGCCAAGA   \n",
       "7                       476.0                 494.0    TTCTGTGCCAGCAGTGAT   \n",
       "9                       350.0                 366.0      TTGTGCCAGCAGCCAA   \n",
       "...                       ...                   ...                   ...   \n",
       "21119                   400.0                 416.0      TGTGCCAGCGGTGATG   \n",
       "21121                   395.0                 411.0      TTCTGTGCCAGCGGTG   \n",
       "21124                   543.0                 562.0   TGTACTTCTGTGCCAGCTC   \n",
       "21126                   399.0                 415.0      TGTGCCAGCGGTGATG   \n",
       "21128                   396.0                 412.0      TGTGCCAGCGGTGATG   \n",
       "\n",
       "       Prim_overlap_start_ind  Prim_overlap_end_ind  \\\n",
       "0                       457.0                 476.0   \n",
       "2                       396.0                 414.0   \n",
       "5                       366.0                 390.0   \n",
       "7                       491.0                 516.0   \n",
       "9                       369.0                 387.0   \n",
       "...                       ...                   ...   \n",
       "21119                   419.0                 439.0   \n",
       "21121                     NaN                   NaN   \n",
       "21124                     NaN                   NaN   \n",
       "21126                   419.0                 436.0   \n",
       "21128                   415.0                 436.0   \n",
       "\n",
       "              Primer_overlap_seq  Primer_overlap_Tm  \\\n",
       "0            GGACTCCGACTACACCTTC          54.489626   \n",
       "2             GGCGGATGGAACACCTTG          56.326643   \n",
       "5       GATAACAATTCTGGAAATACGCTC          54.076533   \n",
       "7      GATAATAGGGGAGAAGTCTTCTTTG          54.571719   \n",
       "9             GGGCAAGACACCCAGTAC          55.201463   \n",
       "...                          ...                ...   \n",
       "21119       GCTGGGGGTTTAACCAAGAC          56.641294   \n",
       "21121                       None                NaN   \n",
       "21124                       None                NaN   \n",
       "21126          GGGACTGGGGGGTCATC          56.161153   \n",
       "21128      GGACTGGGGTTTTTAACCAAG          54.846900   \n",
       "\n",
       "                                  CDR3b_forward  TRBC_overlap_start_ind  \\\n",
       "0      TGCCAGCAGCTTAGCCGGGGGGGACTCCGACTACACCTTC                   504.0   \n",
       "2      GTATCTTTGTGCAAGCAGCTCTGGCGGATGGAACACCTTG                   448.0   \n",
       "5        GTGCCAGCAGCCAAGATAACAATTCTGGAAATACGCTC                   424.0   \n",
       "7      TTCTGTGCCAGCAGTGATAATAGGGGAGAAGTCTTCTTTG                   543.0   \n",
       "9         TTGTGCCAGCAGCCAACTGGGGCAAGACACCCAGTAC                   418.0   \n",
       "...                                         ...                     ...   \n",
       "21119   TGTGCCAGCGGTGATGCGGGCTGGGGGTTTAACCAAGAC                   479.0   \n",
       "21121                                      None                     NaN   \n",
       "21124                                      None                     NaN   \n",
       "21126     TGTGCCAGCGGTGATGCAATGGGACTGGGGGGTCATC                   481.0   \n",
       "21128  TGTGCCAGCGGTGATGCAGGGACTGGGGTTTTTAACCAAG                   478.0   \n",
       "\n",
       "       TRBC_overlap_end_ind       TRBC_overlap_seq  \\\n",
       "0                     525.0  AGGATCTGAGAAATGTGACTC   \n",
       "2                     469.0  AGGATCTGAGAAATGTGACTC   \n",
       "5                     445.0  AGGATCTGAGAAATGTGACTC   \n",
       "7                     564.0  AGGATCTGAGAAATGTGACTC   \n",
       "9                     439.0  AGGATCTGAGAAATGTGACTC   \n",
       "...                     ...                    ...   \n",
       "21119                 500.0  AGGATCTGAGAAATGTGACTC   \n",
       "21121                   NaN                   None   \n",
       "21124                   NaN                   None   \n",
       "21126                 502.0  AGGATCTGAGAAATGTGACTC   \n",
       "21128                 499.0  AGGATCTGAGAAATGTGACTC   \n",
       "\n",
       "                                    CDR3B_Reverse_Primer  \\\n",
       "0      TTTCTCAGATCCTCTATTACCAAAAGCCTGGTCCCTGAGCCGAAGG...   \n",
       "2      CAGATCCTCTAGCACCGATAGTCGGGTGCCCGCACCAAAGTACAAG...   \n",
       "5      CTCTACAACAATGAGCCGGCTTCCTTCTCCAAAATAGAGCGTATTT...   \n",
       "7      CAGATCCTCTACAACTGTGAGTCTGGTTCCTTTACCAAAGAAGACT...   \n",
       "9      TCTCAGATCCTCTAACACGAGGAGCCGAGTGCCTGGCCCAAAGTAC...   \n",
       "...                                                  ...   \n",
       "21119  CTAACACGAGGAGCCGAGTGCCTGGCCCAAAGTACTGGGTGTCTTG...   \n",
       "21121                                               None   \n",
       "21124                                               None   \n",
       "21126  AGCACCGATAGTCGGGTGCCCGCACCAAAGTACAAGGTGTTTTGAT...   \n",
       "21128  ACACGAGGAGCCGAGTGCCTGGCCCAAAGTACTGGGTGTCTTGGTT...   \n",
       "\n",
       "                     CDR3B_Outer_Reverse_Primer  CDR3B_rev_overlap_Tm  \n",
       "0            GAGTCACATTTCTCAGATCCTCTATTACCAAAAG             54.568380  \n",
       "2            GAGTCACATTTCTCAGATCCTCTAGCACCGATAG             54.765029  \n",
       "5       GAGTCACATTTCTCAGATCCTCTACAACAATGAGCCGGC             56.885491  \n",
       "7           GAGTCACATTTCTCAGATCCTCTACAACTGTGAGT             54.796917  \n",
       "9              GAGTCACATTTCTCAGATCCTCTAACACGAGG             56.330476  \n",
       "...                                         ...                   ...  \n",
       "21119   GAGTCACATTTCTCAGATCCTCTAACACGAGGAGCCGAG             54.812132  \n",
       "21121                                      None                   NaN  \n",
       "21124                                      None                   NaN  \n",
       "21126  GAGTCACATTTCTCAGATCCTCTAGCACCGATAGTCGGGT             55.047754  \n",
       "21128  GAGTCACATTTCTCAGATCCTCTAACACGAGGAGCCGAGT             54.787622  \n",
       "\n",
       "[10612 rows x 16 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'''\n",
    "with open(\"TRBVs.fasta.txt\", \"r\") as f:\n",
    "    lines = f.readlines()\n",
    "\n",
    "name = [line.strip() for i, line in enumerate(lines) if i % 2 == 0]\n",
    "trbvs = [line.strip() for i, line in enumerate(lines) if i % 2 == 1]\n",
    "'''\n",
    "def CDRB_reverse(seq, primer_overlap_start, primer_overlap_end, CC):\n",
    "    if pd.isna(primer_overlap_start) or pd.isna(primer_overlap_end):\n",
    "        return None, None, None, None, None, None\n",
    "    primer_overlap_start, primer_overlap_end = int(primer_overlap_start), int(primer_overlap_end)\n",
    "\n",
    "    trbc_overlap_start_ind = seq.find('AGGATCTGAGAAATGTGACTC', primer_overlap_end)\n",
    "    if trbc_overlap_start_ind == -1:\n",
    "        missing('Betas', CC, 'Could not find TRBC')\n",
    "        return None, None, None, None, None, None\n",
    "    \n",
    "    trbc_overlap_end_ind = trbc_overlap_start_ind + len('AGGATCTGAGAAATGTGACTC')\n",
    "    cdr3b_rev_primer_start = primer_overlap_start\n",
    "    cdr3b_rev_primer_end = trbc_overlap_end_ind\n",
    "    while cdr3b_rev_primer_end - cdr3b_rev_primer_start > 60:\n",
    "        cdr3b_rev_primer_end-=1\n",
    "\n",
    "    cdr3b_rev__outer_primer_end = trbc_overlap_end_ind\n",
    "    cdr3b_rev__outer_primer_start = trbc_overlap_start_ind\n",
    "    while pm.calc_tm(seq[cdr3b_rev__outer_primer_start:cdr3b_rev_primer_end]) < 54 or pm.calc_tm(seq[cdr3b_rev__outer_primer_start:cdr3b_rev_primer_end]) > 58 or (not no_repeat_neighbors(seq, cdr3b_rev__outer_primer_start, 'right')):\n",
    "        cdr3b_rev__outer_primer_start-=1\n",
    "        if cdr3b_rev__outer_primer_start < cdr3b_rev_primer_start:\n",
    "            missing('Betas', CC, 'Could not find CDR3B Reverse Outer Primer that worked')\n",
    "            return trbc_overlap_start_ind, trbc_overlap_end_ind, seq[trbc_overlap_start_ind:trbc_overlap_end_ind], rev_comp(seq[cdr3b_rev_primer_start:cdr3b_rev_primer_end]), None, None\n",
    "    \n",
    "    cdr3b_rev_primer = rev_comp(seq[cdr3b_rev_primer_start:cdr3b_rev_primer_end])\n",
    "    cdr3b_rev_out_primer = rev_comp(seq[cdr3b_rev__outer_primer_start:cdr3b_rev__outer_primer_end])\n",
    "    temp_cdr = cdr3b_rev_out_primer\n",
    "    while temp_cdr not in cdr3b_rev_primer:\n",
    "        temp_cdr = temp_cdr[1:]\n",
    "    return trbc_overlap_start_ind, trbc_overlap_end_ind, seq[trbc_overlap_start_ind:trbc_overlap_end_ind], cdr3b_rev_primer, cdr3b_rev_out_primer, pm.calc_tm(temp_cdr)\n",
    "\n",
    "beta = beta.copy()\n",
    "beta[['TRBC_overlap_start_ind', 'TRBC_overlap_end_ind', 'TRBC_overlap_seq', 'CDR3B_Reverse_Primer', 'CDR3B_Outer_Reverse_Primer', 'CDR3B_rev_overlap_Tm']] = beta.apply(\n",
    "    lambda row: pd.Series(CDRB_reverse(row['Seq'], row['Prim_overlap_start_ind'], row['Prim_overlap_end_ind'], row['CC'])),\n",
    "    axis=1\n",
    ")\n",
    "#beta = beta.dropna()\n",
    "beta  \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha.to_csv('Clonotypes_alpha.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "beta.to_csv('Clonotypes_beta.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10522"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(alpha_orig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10612"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(beta_orig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10049"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(alpha.dropna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9429"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(beta.dropna())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
